src/storage.rs

```diff
@@ -2390,23 +2276,359 @@ impl Cache {
+                            info!("recording user info for {api_user:?}");
+                            guard.crates_cache.users.insert(
+                                api_user.id,
+                                CratesCacheUser {
+                                    login: api_user.login,
+                                    name: api_user.name,
+                                },
+                            );
+                            api_user.id
+                        }),
+                    }),
+                )
+            })
+            .collect();
+        info!(
+            "found {} versions for crate {}",
+            versions.len(),
+            self.crate_name
+        );
+        guard.crates_cache.crates.insert(
+            self.crate_name.to_owned(),
+            CratesCacheEntry {
+                last_fetched: self.now,
+                versions,
+                metadata: Some(result.crate_data),
+            },
+        );
+
+        Ok(guard)
+    }
+
+    /// Use `index.crates.io` to get crate information.
+    ///
+    /// This will only add versions which aren't already present in `crates_cache.crates` for
+    /// `name`.
+    ///
+    /// When this function returns `Ok`, the returned state is guaranteed to have an entry for
+    /// `name` in `crates_cache.crates`.
+    async fn update_using_index(
+        &self,
+        network: &Network,
+    ) -> Result<std::sync::MutexGuard<'a, CacheState>, CrateInfoError> {
+        // Crate names can only be a subset of ascii (valid rust identifier characters and `-`), so
+        // using `len()` and indexing will result in valid counts/characters.
+        let mut url = String::from("https://index.crates.io/");
+        let name = self.crate_name;
+        use std::fmt::Write;
+        match name.len() {
+            1 => write!(url, "1/{}", name),
+            2 => write!(url, "2/{}", name),
+            3 => write!(url, "3/{}/{name}", &name[0..1]),
+            _ => write!(url, "{}/{}/{name}", &name[0..2], &name[2..4]),
+        }
+        .expect("writing to a String should not fail");
+        let url = Url::parse(&url).expect("invalid crate name");
+
+        let response = self.try_download(network, url).await?;
+
+        #[derive(Deserialize, Debug, Clone)]
+        struct IndexEntry {
+            vers: semver::Version,
+        }
+
+        let result = load_json_lines::<IndexEntry>(&response[..])?;
+
+        // Update the crates cache with version info (if not already present).
+        let mut guard = self.cache.state.lock().unwrap();
+        info!("found {} versions for crate {}", result.len(), name);
+
+        let entry = guard
+            .crates_cache
+            .crates
+            .entry(name.to_owned())
+            .or_insert_with(|| CratesCacheEntry {
+                last_fetched: self.now,
+                versions: Default::default(),
+                metadata: None,
+            });
+        entry.last_fetched = self.now;
```

Comment:
Given you'll be overriding `last_fetched` anyway, this could probably be `or_default()` I would imagine?

Reply:
I didn't add a `Default` impl because I wasn't sure it made sense semantically to have `last_fetched` have a default. But the logic would all work fine if it happened to be left at that value (assuming `chrono` stuff has defaults that are the beginning of the representable time range).